elasticdl/python/master/optimizer_wrapper.py

```diff
@@ -199,199 +150,71 @@ def _init_thread_local(self):
+                variable.
         """
         # TODO (#1255): Discuss whether `OptimizerWrapper` needs a lock after
         # implementing PS.
         self._init_thread_local()
 
-        grads_and_vars = list(grads_and_vars)
-
-        # split `grads_and_vars` according to whether it is from
-        # ElasticDL embedding layer
-        grads_and_vars_local = []
-        grads_and_vars_kv_store = []
+        grads_and_vars_new = []
         for grad, var in grads_and_vars:
             layer_name = _get_embedding_layer_name_from_var(var)
-            if layer_name:
-                grads_and_vars_kv_store.append((grad, layer_name))
+            if not layer_name:
+                grads_and_vars_new.append((grad, var))
             else:
-                grads_and_vars_local.append((grad, var))
-
-        # `_lookup_embeddings_and_slots` will raise Error if there are
-        # unknown embedding keys
-        embed_values, slot_values = self._lookup_embeddings_and_slots(
-            grads_and_vars_kv_store
-        )
-
-        self._set_embedding_values_to_variables(
-            grads_and_vars_kv_store, embed_values
-        )
-        self._set_slot_values_to_variables(slot_values)
-
-        self._opt.apply_gradients(
-            grads_and_vars_local + grads_and_vars_kv_store
-        )
+                grads_and_vars_new.append(
+                    self._handle_embedding_grad(grad, layer_name)
+                )
 
+        self._opt.apply_gradients(grads_and_vars_new)
         self._report_to_kv_store()
-
         self._delete_variables()
 
-    def _lookup_embeddings_and_slots(self, grads_and_vars):
-        """Look up embedding vectors and slot values form kv store.
-
-        This function looks up embedding vectors and slot values.
-        It initializes unknown slot if exist.
-
-        Arguments:
-            grads_and_vars: A list of (gradient, layer name) pairs.
-
-        Returns:
-            A tuple of (`embedding_values`, `slot_values`). `embedding_values`
-            is a python dictionary of {layer name: `embedding_vectors`} where
-            `embedding_vectors` is a 2D `numpy.ndarray`. `slot_values` is a
-            python dictionary of {layer name: {slot name: `slot_values`}}
-            where `slot_values` is a 2D `numpy.ndarray`.
-
-        Raises:
-            RuntimeError: If any unknown embedding key exists.
-        """
-
-        arr = self._generate_lookup_keys(grads_and_vars)
-        embed_keys, slot_keys, embed_key_index, slot_key_index = arr
-
-        keys = embed_keys + slot_keys
-        embed_keys_num = len(embed_keys)
-        if self._lookup_embedding_func:
-            values, unknown_keys = self._lookup_embedding_func(keys)
-        else:
-            values, unknown_keys = EmbeddingService.lookup_embedding(
-                keys=keys, embedding_service_endpoint=self._kv_store_endpoint
+    def _handle_embedding_grad(self, grad, layer_name):
```

Comment:
`_handle_embedding_grad` to `_get_embedding_var_and_grad`

Reply:
Done